### PR TITLE
Convergence test and validation experiment Buildkite pipeline

### LINKED ISSUE #1216: Run convergence tests on GPU
Might have uncovered issues such as #1170

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR finally sets up a CI pipeline to run convergence tests and validation experiments for CPU and GPU on Buildkite.

Don't think this should run on every push like the main pipeline does and I couldn't figure out how to trigger it via a GitHub comment (see https://github.com/CliMA/slurm-buildkite/issues/13).

We can trigger this pipeline manually from Buildkite and I've scheduled it to run every night at 3am EST (on the master branch).

Cool thing is that it uploads the convergence plots as artifacts so we can view them from Buildkite!

image

Resolves #1216

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
Quite concerning that some of the convergence tests do not pass on the GPU...

This is probably due to #1170 since the simpler convergence tests that do not rely on a pressure solver seem to pass (also the fact that it passes forced flow free slip with doubly periodic (x, z) but not with a wall-bounded dimension (x, y)).

image

image

**Review by glwagner — APPROVED** [review]
This is great. Some minor comments:

* `test/test_convergence.jl` is rather boiler-plate-y. Is there any way to use a list of convergence tests scripts? This will make it easier to add new tests in the future...

* Could make sense to raise an issue about converting convergence tests plots to `Plots.jl` or `CairoMakie.jl` or whatever when this PR is merged, so we don't have to wrestle `PyPlot` in the test scripts.

**Comment by ali-ramadhan** [comment]
>  * `test/test_convergence.jl` is rather boiler-plate-y. Is there any way to use a list of convergence tests scripts? This will make it easier to add new tests in the future...

Yeah it could be more automated. I'll give it a try once the tests are all passing.

>  * Could make sense to raise an issue about converting convergence tests plots to `Plots.jl` or `CairoMakie.jl` or whatever when this PR is merged, so we don't have to wrestle `PyPlot` in the test scripts.

I don't think it's worth spending much effort on switching plotting libraries (especially since the plots work now) but I'll use CairoMakie.jl for future plots. CairoMakie.jl doesn't support log axes yet (gotta hack it in for now) but hopefully it'll get added soon.

**Comment by navidcy** [comment]
Some tests fail..

**Comment by ali-ramadhan** [comment]
All tests should pass (see https://buildkite.com/clima/oceananigans-validation-experiments/builds/96) but Tartarus went down so a lot of builds died leading to failing tests.

File: validation/convergence_tests/one_dimensional_advection_schemes.jl [hunk 1]
```diff
@@ -163,14 +163,15 @@ for scheme in advection_schemes
 
         xscale("log", base=2)
         yscale("log", base=10)
+        title("Convergence for $name advection scheme")
         xlabel(L"N_x")
         ylabel("\$L\$-norms of \$ | c_\\mathrm{sim} - c_\\mathrm{analytical} |\$")
         removespines("top", "right")
-        legend = legend(loc="upper right", bbox_to_anchor=(1.4, 1.0), prop=Dict(:size=>6))
+        lgd = legend(loc="upper right", bbox_to_anchor=(1.4, 1.0), prop=Dict(:size=>6))
```
Comment (glwagner) [inline validation/convergence_tests/one_dimensional_advection_schemes.jl:170]: 👁️

**Comment by ali-ramadhan** [comment]
44 successful checks :rocket:

I think this PR is ready to be merged. All the existing convergence tests are now in CI and pass, and the plots are uploaded as Buildkite artifacts. This PR might conflict with @francispoulin's PR #1276 (I'm happy to fix conflicts no matter which PR is merged first).

There are a few tests that take a long time. This can be shortened in a future PR as it might involve some trial and error and fiddling with rate of convergence tolerances. Other validation experiments should also be added.

Right now I have set the validation pipeline to only run on master (and every night at 3am ET).

**Comment by francispoulin** [comment]
> 44 successful checks 
> 
> I think this PR is ready to be merged. All the existing convergence tests are now in CI and pass, and the plots are uploaded as Buildkite artifacts. This PR might conflict with @francispoulin's PR #1276 (I'm happy to fix conflicts no matter which PR is merged first).
> 
> There are a few tests that take a long time. This can be shortened in a future PR as it might involve some trial and error and fiddling with rate of convergence tolerances. Other validation experiments should also be added.
> 
> Right now I have set the validation pipeline to only run on master (and every night at 3am ET).

Please go ahead and don't wait for me.  I will work around whatever you have done here in my PR.

**Pull request merged**